### ISSUE TITLE
Timeout is reset on stream when calling `into_reader()`

### ISSUE THREAD

**Issue by birkenfeld** [algesten/ureq#196]
When calling `into_reader()` on a request, the timeout set with `set_timeout_read` (NOT the deadline `set_timeout`) on the request is reset to no timeout by `unit::connect()` calling `DeadlineStream::reset` before the socket is returned, and there is no way to set it again.

This makes `set_timeout_read` basically unused except for reading the headers.

**Comment by birkenfeld** [comment]
In fact, it seems that `timeout_read` also has a bug when a read timeout appears in the header part.

Below is a reproducing example (using version 1.5.1 here).

```
use std::{net, thread, time, io::prelude::*};

fn main() {
    thread::spawn(server);
    thread::sleep(time::Duration::from_millis(100));

    let x = ureq::get("http://localhost:12345/").timeout_read(100).call();
    println!("Response string: {:?}", x.into_string());
}

const WAIT_HEAD: time::Duration = time::Duration::from_millis(0);
const WAIT_BODY: time::Duration = time::Duration::from_millis(0);

fn server() {
    let listener = net::TcpListener::bind(("localhost", 12345)).unwrap();
    let (mut sock, _) = listener.accept().unwrap();
    let mut buf = [0; 128];
    sock.read(&mut buf).unwrap();

    thread::sleep(WAIT_HEAD);
    sock.write_all(b"HTTP/1.1 200 OK\r\nContent-Length: 10\r\n").unwrap();
    sock.write_all(b"Connection: close\r\nContent-Type: text/html\r\n\r\n").unwrap();

    thread::sleep(WAIT_BODY);
    sock.write_all(b"xxxxxxxxxx").unwrap();
}
```

If this is run as given, with no sleeps, everything looks fine:
```
Response string: Ok("xxxxxxxxxx")
```

If run with `WAIT_BODY` set to 1000 milliseconds:
```
Response string: Ok("xxxxxxxxxx")
```
instead of a timeout error.

If run with `WAIT_HEAD` set to 1000 milliseconds:
```
Response string: Ok("HTTP/1.1 200 OK\r\nContent-Length: 10\r\nConnection: close\r\nContent-Type: text/html\r\n\r\nxxxxxxxxxx")
```

**Comment by jsha** [comment]
Thanks for the report, and the functional repro case! I'm looking into it.

**Comment by jsha** [comment]
Alright, I think I've figured this out. First, if you'd like to work around this for now, you can check `x.ok()` before calling `into_string()`; in my tests with your repro case, `x.ok()` is returning false, because this is getting a timeout, and that's being turned into a "synthetic error" response. Also: we're actively working on a 2.0 release that will get rid of the synthetic error concept and use more familiar Result-based handling throughout.

That said, this is still a bug - if the Response is a synthetic error, `into_string()` should return an error. That used to be the case. Using `git bisect`, I found that I broke it in #67.

Specifically, in unit.rs, I added a DeadlineStream wrapping the main TcpStream: https://github.com/algesten/ureq/pull/67/files#diff-3f95f2fa03d78724ed201d3c200b853a1652649971512b351d7a9d2a26d8a637R168.

When the timeout hits, DeadlineStream returns an error. That happens inside `Response.from_read`: https://github.com/algesten/ureq/blob/57be414d9758bd1bc5f607645cd21666c9f92a9a/src/response.rs#L444-L446. from_read catches the error and turns it into a Response containing a synthetic error. That will be the Response that is returned.

However, since from_read took a `&mut` reference to the DeadlineStream and did not consume it, the stream can be reused further down in unit::connect: https://github.com/algesten/ureq/commit/57be414d9758bd1bc5f607645cd21666c9f92a9a#diff-3f95f2fa03d78724ed201d3c200b853a1652649971512b351d7a9d2a26d8a637R224

```
    crate::response::set_stream(&mut resp, unit.url.to_string(), Some(unit), stream.into());
```

That `stream.into()` unwraps the inner stream from the DeadlineStream, resetting all the timeouts. Now `resp` is a synthetic error response that happens to contain a valid stream with no timeouts, which can be consumed by `into_string()`.

I believe the quick fix for this is to check `resp.synthetic_error()` immediately after calling `Response::from_read` in `connect(...)`, and return early if there's a synthetic error. I'll work on a patch for that. I think for the 2.0 branch this is already fixed by more thorough handling of errors - `Response.from_read` return a `Result<Response, Error>` instead of a `Response`.

Also, looking at this code again, it appears that `into_reader()` doesn't obey timeout_read, only deadline. I believe this is the source of the bug you mentioned in your first comment. Prior to #67, I think timeouts were set on streams and never cleared. This worked fine for the no-agent case, since streams were dropped rather than pooled; but it would have always been broken in the agent case.

**Comment by birkenfeld** [comment]
@jsha Thanks very much for the analysis!  I didn't think to check for the synthetic error, but it explains why the headers are part of the body - of course one can't really rely on the body being correct in these cases.

My actual use case is the one from the first comment - where the timeout condition is likely to appear during reading the body.  My reason for using `timeout_read` instead of `timeout` is that I'm reading (and processing) quite big streams of data. The bottleneck is on the processing side, so an overall deadline for the HTTP side is quite hard to select.  However, the read timeout handles cases of server fails, or severed connections (e.g. ethernet cable pulled out) well since it applies to individual socket reads.  Do you think this could be fixed in the 1.5 series?

**Comment by jsha** [comment]
Yes, I think the first bug (read timeout is not applied during into_string) also makes sense to fix in the 1.5 series. I'm pretty sure it is also present in the 2.0 branch and will need to be fixed there too. If we make a patch against 1.5 I think it will probably port forward pretty straightforwardly.
